doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,322 @@
+requirements on the use of documentation—e.g., a policy that requires that every
+field in every schema published to the organization's schema registry has
+documentation.
+
+## Goals
+
+Must haves:
+
+  * Enable users with organizational policies on the use of `doc` fields in Avro
+    schemas to use Avro-formatted sinks.
+
+Nice to haves:
+
+  * Consistency with Materialize's existing approach to generating Avro schemas
+    from SQL relations.
+
+  * Good ergonomics.
+
+## Non-goals
+
+None yet.
+
+## Detailed description
+
+### Design
+
+#### SQL
+
+We'll extend the `FORMAT AVRO` clause with a new `DOC ON` option:
+
+```sql
+CREATE SINK ...
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY ... (
+    DOC ON <field-specifier> = '<documentation string>',
+    DOC ON <field-specifier> = '<documentation string>',
+    ...
+)
+WITH (...)
+```
+
+The option will be located in a new parenthesized option list immediately
+following the existing `FORMAT AVRO` clause. After this change, `DOC ON` will
+be the only supported option, but we expect to add further options to control
+the Avro schema in the future.
+
+Aside: the current design of the `USING CONFLUENT SCHEMA REGISTRY` option is
+historical and predates the introduction of the standardized option list. In the
+future, we may also move the `USING CONFLUENT SCHEMA REGISTRY` into the option
+list, where it can be freely intermixed with other options, e.g.:
+
+```sql
+FORMAT AVRO (
+    DOC ON ...,
+    CONFLUENT SCHEMA REGISTRY = '...',
+    DOC ON ...,
+)
+```
+
+Each `DOC ON` option describes a documentation comment that should be attached
+to a single field in the Avro schema. See [Avro schema node
+specifiers](#avro-schema-node-specifiers) below for details on how users name
+the Avro schema node on which to attach documentation.
+
+Materialize will reject `CREATE SINK` commands that specify multiple `DOC ON`
+options naming the same field.
+
+Materialize will automatically use [comments] on the columns of materialized
+views, or its constituent types, as the default documentation for a field.
+Specifically, for a given node of the Avro schema, Materialize will consider the
+following sources of documentation in order:
+
+  * A `DOC ON` option specifically naming that schema node.
+  * A comment on the column that produced the schema node. If the schema
+    element is at the top level, this will be a comment on a column of the
+    sink's materialized view. If the schema node is nested, this will be a
+    comment on a column of a record type.
+  * A comment on the type of the column that produced the schema node.
+
+For example, consider the following sink:
+
+```sql
+CREATE TYPE point AS (x integer, y integer);
+CREATE MATERIALIZED VIEW v AS SELECT ROW(1, 1)::point AS c1, 'text' AS c2;
+CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO ...;
+```
+
+For `c1`, Materialize will consider the following sources of documentation
+in order:
+
+  * A `DOC ON` option naming `c1`.
```

Comment:
So, if the sink is of type envelope upsert, the full name will be `envelope.c1` if it's in the value and `row.c1` if it's in key [link](https://materialize.com/docs/sql/create-sink/kafka/#csr-connection-options)?

Reply:
Oh gosh, good call out. I really wish we'd named them both `envelope`. Darn! Maybe we want separate options? `KEY DOC ON` and `VALUE DOC ON`? Not sure. Need to noodle more.

Reply:
Would that be too bad if we change them to `envelope`? From what it looks like after we create the schema, we store the schema ID and keep referring to it, so should still work fine?

Reply:
`KEY DOC ON` and `VALUE DOC ON` should work. Also, I think it's better to have separate options. Since the key and the value are separate schema objects, they could potentially have same record names?

We'll have to be thorough with covering and testing all the sink envelope combinations.

Reply:
I think we'll probably want all three? I think the common case is sharing documentation between the key and value schemas, in which case you want the ergonomics of `DOC ON`.